app/service/planning_svc.py

```diff
@@ -91,10 +92,12 @@ def __init__(self):
                 agent_facts = await self._get_agent_facts(operation['id'], agent['paw'])
                 relevant_facts = await self._build_relevant_facts(variables, operation.get('facts', []), agent_facts)
                 valid_facts = await RuleSet(rules=operation.get('rules', [])).apply_rules(facts=relevant_facts[0])
+                ability_requirements = await self._get_operation_requirements(link, operation)
```

Comment:
we should consider calculating ability_requirements outside of the 'for link in links' loop, then we can look up a requirement while we're inside the loop, instead of looping through multiple times.